### ISSUE TITLE
TOC generation: Incorrect escaping in no-utf8 mode

### ISSUE THREAD

**Issue by ntrrgc** [cdrdao/cdrdao#32]
In `CdTextItem::print()`:

```cpp
for (auto c : data_) {
    if (c >= 128)
        out << "\\" << std::oct << (unsigned int)c;
    else
        out << c;
}
```

Because the above code doesn't escape the backslash character, **it is impossible for a reader** to know if a backslash means "here follows a sequence of octal digits representing a byte" or "here is the backslash character".

There are two typical approaches to fix this, either encode `\` as `\\` or as `\134` (134 being the octal representation for the `\` character).

**Comment by ntrrgc** [comment]
This is the problem behind a number of ripping errors in whipper, like https://github.com/whipper-team/whipper/issues/183

**Comment by ntrrgc** [comment]
Oh! While looking at an older version of the code I've also realized the current version also doesn't handle the quotes characters, which 1.2.2 did.

```cpp
      if (data_[i] == '"') {
	out << "\\\"";
      }
      else if (isprint(data_[i])) {
	out << data_[i];
      }
      else {
	sprintf(buf, "\\%03o", (unsigned int)data_[i]);
	out << buf;
      }
```

**Comment by ntrrgc** [comment]
And the UTF-8 mode doesn't escape quote characters either!

```cpp
out << " \"" << u8text << "\"";
```

**Comment by ntrrgc** [comment]
It keeps getting worse. The current code is not handling non-printable ASCII characters, including the NUL byte.

**Comment by denisleroy** [comment]
I understand the issue. There is indeed nothing in the TOC format that would specify whether the TOC file contains UTF-8 for the simple reason that there is no equivalent in the CDDA specification. Essentially, `file foo.toc` will actually tell you.

There is no ambiguity: the TOC file either is pure ASCII, or has some UTF-8 in certain strings, but the data actually burned onto the CD is going to be the same.

Example:
```
  LANGUAGE 0 {
    ENCODING_ISO_8859_1
    TITLE "Debug"
    PERFORMER "Björk"
  }
```

and

```
  LANGUAGE 0 {
    ENCODING_ISO_8859_1
    TITLE "Debug"
    PERFORMER "Bj/316rk"     (-> ISO-8859-1 encoding)
  }
```

Both TOCs will burn the same content, as cdrdao will convert the UTF-8 into ISO-8859-1 as best it can. We're talking about input TOC files here. This is really just a convenience.

As far as generating TOC files however, currently the default is to use UTF8 to make things more readable. I can see how this could break some other tools that rely on cdrdao to generate a TOC file for them. Should the `--no-utf8` behavior be the default and rather have a `--utf8` option ?

**Comment by ntrrgc** [comment]
Right. Any non-ASCII character that is not part of UTF-8 would be escaped anyway.

Regarding whether UTF8 or no-UTF8 should be the default... I'm not sure. If we were starting from a clean slate, my preference would be clearly UTF8, because it's easier to handle without mistakes. I think currently whipper can't handle MSJIS in no-UTF8 mode, but I haven't actually tested it.

What applications do you know that handle the same .toc files with CD-Text cdrdao generates? It would be particularly interesting to see how they handle Latin-1 and MSJIS in both modes.

**Comment by denisleroy** [comment]
> What applications do you know that handle the same .toc files with CD-Text cdrdao generates? It would be particularly interesting to see how they handle Latin-1 and MSJIS in both modes.

CDemu for example accepts a TOC file as input, but won't like seeing UTF-8 code in there. It won't complain, but it won't encode it correctly internally.
